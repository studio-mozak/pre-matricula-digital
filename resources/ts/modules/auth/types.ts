export interface Auth {
  level: number;
  name: string;
  schools: string[];
}
