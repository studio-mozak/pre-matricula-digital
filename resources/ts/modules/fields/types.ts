import { Nullable } from '@/types';

export type SetValues = (fields: Record<string, unknown>) => void;

export interface FieldPageOption {
  id?: string;
  name: Nullable<string>;
  weight: number;
}

export interface FieldPage {
  id: Nullable<string>;
  name: Nullable<string>;
  description: Nullable<string>;
  group: Nullable<'RESPONSIBLE' | 'STUDENT'>;
  type: Nullable<string>;
  internal: boolean;
  options: FieldPageOption[];
  deleteOptions?: string[];
  required?: boolean;
}

export interface ErrorResponse {
  message: string;
}
