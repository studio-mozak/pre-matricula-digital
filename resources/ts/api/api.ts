import axios, { AxiosInstance } from 'axios';

const token = window.config ? window.config.token : '';

export const http: AxiosInstance = axios.create({
  baseURL: String(import.meta.env.VITE_APP_API_URL || ''),
  headers: {
    Authorization: `Token ${token}`,
  },
});

axios.interceptors.response.use(
  (res) => {
    /**
     * If the response from the server has the "errors" object or the object "data" is empty,
     * here it is forced to go to the catch block automatically.
     */
    if (
      (res.data.errors || Object.keys(res.data).length === 0) &&
      !(res.data instanceof Blob)
    ) {
      return Promise.reject({
        response: {
          data: res.data,
        },
      });
    }

    return res;
  },
  (err) => Promise.reject(err)
);

export const rest = <T>(
  type: 'get' | 'post' | 'put' | 'delete',
  url: string,
  params?: object
): Promise<T> => http[type](url, params);

export const graphql = <T>(data: object, config?: object): Promise<T> =>
  axios.post('/graphql', data, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
    ...config,
  });
