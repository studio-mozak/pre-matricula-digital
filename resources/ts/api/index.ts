export * from './api';
export * as City from './services/graphql/city';
export * as Resource from './services/graphql/resource';
