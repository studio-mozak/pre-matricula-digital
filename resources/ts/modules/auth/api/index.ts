/**
 * Exporting the Axios Instance
 */
export { http, graphql } from '@/api';

/**
 * Exporting all modules
 */
export * as AuthRest from '@/modules/auth/api/services/rest/auth';
