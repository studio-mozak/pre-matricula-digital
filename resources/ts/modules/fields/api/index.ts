/**
 * Exporting the Axios Instance
 */
export { http, graphql } from '@/api';

/**
 * Exporting all modules
 */
export * as Field from './services/graphql/field';
