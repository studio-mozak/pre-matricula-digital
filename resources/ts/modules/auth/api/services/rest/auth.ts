import { Auth } from '@/modules/auth/types';
import { http } from '@/modules/auth/api';

export const getAuth = (): Promise<Auth> =>
  http
    .get('/auth/login', {
      withCredentials: true,
    })
    .then((res) => res.data);
