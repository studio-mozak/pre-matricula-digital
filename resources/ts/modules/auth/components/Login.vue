<template>
  <main>
    <div class="text-center" v-html="message"></div>
  </main>
</template>

<script setup lang="ts">
import { Auth } from '@/modules/auth/types';
import { AuthRest } from '@/modules/auth/api';
import { ref } from 'vue';
import { useLoaderAndThrowError } from '@/composables';
import { useRouter } from 'vue-router';

const router = useRouter();

const message = ref<string>('Logando..');

const { loader } = useLoaderAndThrowError<Auth>();

loader(AuthRest.getAuth)
  .then((res) => {
    if (!res) {
      throw new Error();
    }

    router.push({
      name: 'preregistrations',
    });
  })
  .catch(() => {
    message.value = `
      O login de acesso no <strong>Pré-matrícula Digital</strong> é exclusivo para secretários e diretores escolares,
      ou gestores da Secretaria de Educação.
      <br>
      No momento, para acessar o sistema você deve estar logado no Ensinus.
    `;
  });
</script>
