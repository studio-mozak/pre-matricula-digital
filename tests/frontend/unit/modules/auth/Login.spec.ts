import { VueWrapper, flushPromises, shallowMount } from '@vue/test-utils';
import { Auth } from '@/modules/auth/types';
import { AuthRest } from '@/modules/auth/api';
import Login from '@/modules/auth/components/Login.vue';
import { expect, vi } from 'vitest';

vi.mock('@/modules/auth/api');

const spyPush = vi.fn();

vi.mock('vue-router/dist/vue-router.mjs', () => ({
  useRouter: () => ({
    push: spyPush,
  }),
}));

const loginMock: Auth = {
  level: 1,
  name: 'ESCOLA ESTADUAL DE CIENCIAS DA SAÚDE',
  schools: [],
};

const errorMessage = `
      O login de acesso no <strong>Pré-matrícula Digital</strong> é exclusivo para secretários e diretores escolares,
      ou gestores da Secretaria de Educação.
      <br>
      No momento, para acessar o sistema você deve estar logado no i-Educar.
`;

const onMounted = async (success: boolean) => {
  if (success) {
    vi.mocked(AuthRest.getAuth).mockResolvedValueOnce(loginMock);
  } else {
    vi.mocked(AuthRest.getAuth).mockRejectedValueOnce(null);
  }

  const wrapper = shallowMount(Login);

  await flushPromises();

  return wrapper;
};

describe('Login', () => {
  let wrapper: VueWrapper<InstanceType<typeof Login>>;

  beforeEach(async () => {
    wrapper = await onMounted(true);

    await wrapper.vm.$nextTick();
  });

  afterEach(() => {
    vi.clearAllMocks();
    vi.resetAllMocks();
  });

  test('the component exists', async () => {
    expect(wrapper).toBeTruthy();
  });

  test('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('login has made successfully', async () => {
    expect(spyPush).toHaveBeenCalledTimes(1);
    expect(spyPush).toHaveBeenCalledWith({ name: 'preregistrations' });
  });

  test('displays error message if success is false', async () => {
    wrapper = await onMounted(false);

    await wrapper.vm.$nextTick();

    expect(spyPush).toHaveBeenCalledTimes(1);
    expect(wrapper.vm.message).toContain(errorMessage);
  });
});
