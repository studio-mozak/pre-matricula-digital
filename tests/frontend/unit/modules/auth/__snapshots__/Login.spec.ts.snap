// Vitest Snapshot v1

exports[`Login > should match snapshot 1`] = `
<main>
  <div
    class="text-center"
  >
    Logando..
  </div>
</main>
`;
